ksp_plugin/plugin.cpp

```diff
@@ -616,9 +646,112 @@ Velocity<World> Plugin::BubbleVelocityCorrection(
+      CHECK(it != celestial_to_index.end());
+      Index const parent_index = it->second;
+      celestial_message->set_parent_index(parent_index);
+    }
+  }
+  std::map<not_null<Vessel const*>, GUID const> vessel_to_guid;
+  for (auto const& guid_vessel : vessels_) {
+    std::string const& guid = guid_vessel.first;
+    not_null<Vessel*> const vessel = guid_vessel.second.get();
+    vessel_to_guid.emplace(vessel, guid);
+    auto const vessel_message = message->add_vessel();
+    vessel_message->set_guid(guid);
+    vessel->WriteToMessage(vessel_message->mutable_vessel());
+    auto const it = celestial_to_index.find(&vessel->parent());
+    CHECK(it != celestial_to_index.end());
+    Index const parent_index = it->second;
+    vessel_message->set_parent_index(parent_index);
+    vessel_message->set_dirty(is_dirty(vessel));
+  }
+  bubble_->WriteToMessage(
+      [&vessel_to_guid](not_null<Vessel const*> const vessel) -> GUID {
+        auto const it = vessel_to_guid.find(vessel);
+        CHECK(it != vessel_to_guid.end());
+        return it->second;
+      },
+      message->mutable_bubble());
+  planetarium_rotation_.WriteToMessage(message->mutable_planetarium_rotation());
+  current_time_.WriteToMessage(message->mutable_current_time());
+  auto const it = celestial_to_index.find(sun_);
+  CHECK(it != celestial_to_index.end());
+  Index const sun_index = it->second;
+  message->set_sun_index(sun_index);
+}
+
+std::unique_ptr<Plugin> Plugin::ReadFromMessage(
+    serialization::Plugin const& message) {
+  IndexToOwnedCelestial celestials;
+  for (auto const& celestial_message : message.celestial()) {
+    celestials.emplace(
+        celestial_message.index(),
+        Celestial::ReadFromMessage(celestial_message.celestial()));
+  }
+  for (auto const& celestial_message : message.celestial()) {
+    if (celestial_message.has_parent_index()) {
+      auto const it = celestials.find(celestial_message.index());
+      CHECK(it != celestials.end());
+      not_null<std::unique_ptr<Celestial>> const& celestial = it->second;
+      auto const parent_it = celestials.find(celestial_message.parent_index());
+      CHECK(parent_it != celestials.end());
+      not_null<Celestial const*> const parent = parent_it->second.get();
+      celestial->set_parent(parent);
+    }
+  }
+  GUIDToOwnedVessel vessels;
+  std::set<not_null<Vessel*> const> dirty_vessels;
+  for (auto const& vessel_message : message.vessel()) {
+    auto const parent_it = celestials.find(vessel_message.parent_index());
+    CHECK(parent_it != celestials.end());
+    not_null<Celestial const*> const parent = parent_it->second.get();
+    auto const inserted_vessel =
+        vessels.emplace(vessel_message.guid(),
+                        Vessel::ReadFromMessage(vessel_message.vessel(),
+                                                parent));
+    not_null<Vessel*> const vessel = inserted_vessel.first->second.get();
```

Comment:
first->second is totally incomprehensible.  Instead: (1) call ReadFromMessage, result in a local unique ptr; (2) call get() to make the vessel dirty; (3) std::move the unique_ptr into the map.